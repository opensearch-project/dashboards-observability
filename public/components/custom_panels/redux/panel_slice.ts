/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { createSelector, createSlice } from '@reduxjs/toolkit';
import { concat, from, Observable, of } from 'rxjs';
import { map, mergeMap, toArray } from 'rxjs/operators';
import { forEach } from 'lodash';
import {
  createDemoPanel,
  CUSTOM_PANELS_API_PREFIX,
  CUSTOM_PANELS_SAVED_OBJECT_TYPE,
  samplePanelName,
} from '../../../../common/constants/custom_panels';
import {
  CustomPanelListType,
  CustomPanelType,
  ObservabilityPanelAttrs,
  PanelType,
} from '../../../../common/types/custom_panels';
import { coreRefs } from '../../../framework/core_refs';
import { SavedObject, SimpleSavedObject } from '../../../../../../src/core/public';
import { addVisualizationPanel } from '../helpers/add_visualization_helper';

interface InitialState {
  id: string;
  panel: CustomPanelType;
  panelList: CustomPanelType[];
}

export const newPanelTemplate = (newName): PanelType => ({
  title: newName,
  dateCreated: new Date().getTime(),
  dateModified: new Date().getTime(),
  visualizations: [],
  queryFilter: { language: '', query: '' },
  timeRange: { from: 'now-1d', to: 'now' },
});

const initialState: InitialState = {
  id: '',
  panel: newPanelTemplate(''),
  panelList: [],
  loadingFlag: false,
};

export const panelSlice = createSlice({
  name: 'customPanel',
  initialState,
  reducers: {
    setPanelId: (state, action) => ({ ...state, id: action.payload }),

    setPanel: (state, action) => {
      return { ...state, panel: action.payload };
    },

    setPanelList: (state, action) => {
      return { ...state, panelList: action.payload };
    },
  },
});

export const { setPanel, setPanelList } = panelSlice.actions;

export const panelReducer = panelSlice.reducer;

export const selectPanel = createSelector(
  (rootState) => rootState.customPanel.panel,
  (panel) => normalizedPanel(panel)
);

const normalizedPanel = (panel: CustomPanelType): CustomPanelType => ({
  ...newPanelTemplate(''),
  ...panel,
});

export const selectPanelList = (rootState): CustomPanelType[] => rootState.customPanel.panelList;

/*
 ** ASYNC DISPATCH FUNCTIONS
 */

const fetchSavedObjectPanels$ = () =>
  from(savedObjectPanelsClient.find()).pipe(
    mergeMap((res) => res.savedObjects),
    map(savedObjectToCustomPanel)
  );

const fetchObservabilityPanels$ = () =>
  of(coreRefs.http.get(`${CUSTOM_PANELS_API_PREFIX}/panels`)).pipe(
    mergeMap((res) => res),
    mergeMap((res) => res.panels as ObservabilityPanelAttrs[]),
    map((p: ObservabilityPanelAttrs) => ({
      ...p,
      title: p.name,
      savedObject: false,
      type: 'observability-savedObject',
    }))
  );

// Fetches all saved Custom Panels
const fetchCustomPanels = async () => {
  const panels$: Observable<CustomPanelListType> = concat(
    fetchSavedObjectPanels$(),
    fetchObservabilityPanels$()
  ).pipe(
    map((res) => {
      return res as CustomPanelListType;
    })
  );

  return panels$.pipe(toArray()).toPromise();
};

export const fetchPanels = () => async (dispatch, getState) => {
  const panels = await fetchCustomPanels();
  dispatch(setPanelList(panels));
};

export const fetchPanel = (id) => async (dispatch, getState) => {
  const soPanel = await savedObjectPanelsClient.get(id);
  const panel = savedObjectToCustomPanel(soPanel);
  dispatch(setPanel(panel));
};

export const fetchVisualization = () => (dispatch, getState) => {};

const updateLegacyPanel = (panel: CustomPanelType) =>
  coreRefs.http!.post(`${CUSTOM_PANELS_API_PREFIX}/panels/update`, {
    body: JSON.stringify({ panelId: panel.id, panel: panel as PanelType }),
  });

const updateSavedObjectPanel = (panel: CustomPanelType) => savedObjectPanelsClient.update(panel);

export const uuidRx = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/;

export const isUuid = (id) => !!id.match(uuidRx);

export const updatePanel = (
  panel: CustomPanelType,
  successMsg: string,
  failureMsg: string
) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  try {
    if (isUuid(panel.id)) await updateSavedObjectPanel(panel);
    else await updateLegacyPanel(panel);
    if (successMsg) {
      toasts!.add(successMsg);
    }
    dispatch(setPanel(panel));
    const panelList = getState().customPanel.panelList.map((p) => (p.id === panel.id ? panel : p));
    dispatch(setPanelList(panelList));
  } catch (e) {
    if (failureMsg) {
      toasts!.addDanger(failureMsg);
    }
    console.error(e);
  }
};

export const addVizToPanels = (panels, vizId) => async (dispatch, getState) => {
  forEach(panels, (oldPanel) => {
    const panel = getState().customPanel.panelList.find((p) => p.id === oldPanel.panel.id);

    const allVisualizations = panel!.visualizations;

    const visualizationsWithNewPanel = addVisualizationPanel(vizId, undefined, allVisualizations);

    const updatedPanel = { ...panel, visualizations: visualizationsWithNewPanel };
    dispatch(updatePanel(updatedPanel, '', ''));
  });
};

export const addMultipleVizToPanels = async (panels, vizIds) => {
  await Promise.all(
    panels.map(async ({ panel: { id } }) => {
      const soPanel = await savedObjectPanelsClient.get(id);
      const oldPanel = savedObjectToCustomPanel(soPanel);

      const allVisualizations = oldPanel!.visualizations;

      let visualizationsWithNewPanel = allVisualizations;

      forEach(vizIds, (vizId) => {
        visualizationsWithNewPanel = addVisualizationPanel(
          vizId,
          undefined,
          visualizationsWithNewPanel
        );
      });

      const updatedPanel = { ...oldPanel, visualizations: visualizationsWithNewPanel };

      if (isUuid(updatedPanel.id)) {
        await updateSavedObjectPanel(updatedPanel);
      } else {
        await updateLegacyPanel(updatedPanel);
      }
    })
  );
};

export const replaceVizInPanel = (oldPanel, oldVizId, vizId, newVisualizationTitle) => async (
  dispatch,
  getState
) => {
  const panel = getState().customPanel.panelList.find((p) => p.id === oldPanel.id);

  const allVisualizations = panel!.visualizations;

  const visualizationsWithNewPanel = addVisualizationPanel(vizId, oldVizId, allVisualizations);

  const updatedPanel = { ...panel, visualizations: visualizationsWithNewPanel };

  dispatch(
    updatePanel(
      updatedPanel,
      `Visualization ${newVisualizationTitle} successfully added!`,
      `Error in adding ${newVisualizationTitle} visualization to the panel`
    )
  );
};

const deletePanelSO = (customPanelIdList: string[]) => {
  const soPanelIds = customPanelIdList.filter((id) => isUuid(id));
  return Promise.all(soPanelIds.map((id) => savedObjectPanelsClient.delete(id)));
};

const deleteLegacyPanels = (customPanelIdList: string[]) => {
  const panelIds = customPanelIdList.filter((id) => !isUuid(id));
  if (panelIds.length === 0) return;

  const concatList = panelIds.toString();
  return coreRefs.http!.delete(`${CUSTOM_PANELS_API_PREFIX}/panelList/` + concatList);
};

export const deletePanels = (panelsToDelete: CustomPanelType[]) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  const toastMessage = `Observability Dashboard${
    panelsToDelete.length > 1 ? 's' : ' ' + panelsToDelete[0].title
  } successfully deleted!`;
  try {
    const ids = panelsToDelete.map((p) => p.id);
    await Promise.all([deleteLegacyPanels(ids), deletePanelSO(ids)]);

    const panelList: CustomPanelType[] = getState().customPanel.panelList.filter(
      (p) => !ids.includes(p.id)
    );
    dispatch(setPanelList(panelList));
    toasts!.add(toastMessage);
  } catch (e) {
    toasts!.addDanger(
      'Error deleting Observability Dashboards, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

export const createPanel = (panel) => async (dispatch, getState) => {
  const { toasts } = coreRefs;
  try {
    const newSOPanel = await savedObjectPanelsClient.create(panel);
    const newPanel = savedObjectToCustomPanel(newSOPanel);
    const panelList = getState().customPanel.panelList;
    dispatch(setPanelList([...panelList, newPanel]));
    toasts!.add(`Observability Dashboard "${newPanel.title}" successfully created!`);
    window.location.replace(`#/${newPanel.id}`);
  } catch (e) {
    toasts!.addDanger(
      'Error occurred while creating Observability Dashboard, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

export const createPanelSample = (vizIds) => async (dispatch, getState) => {
  const samplePanel = {
    ...createDemoPanel(vizIds),
    dateCreated: new Date().getTime(),
    dateModified: new Date().getTime(),
    title: samplePanelName,
  };
  const newSOPanel = await savedObjectPanelsClient.create(samplePanel);
  const newPanel = savedObjectToCustomPanel(newSOPanel);
  const panelList = getState().customPanel.panelList;
  dispatch(setPanelList([...panelList, newPanel]));
};

export const clonePanel = (panel, newPanelName) => async (dispatch, getState) => {
  const { toasts } = coreRefs;

  try {
    const { id, ...panelCopy } = {
      ...panel,
      title: newPanelName,
      dateCreated: new Date().getTime(),
      dateModified: new Date().getTime(),
    } as PanelType;

    const newSOPanel = await savedObjectPanelsClient.create(panelCopy);

    const newPanel = savedObjectToCustomPanel(newSOPanel);
    const panelList = getState().customPanel.panelList;
    dispatch(setPanelList([...panelList, newPanel]));
    dispatch(setPanel(newPanel));
    toasts!.add(`Observability Dashboard "${newPanel.title}" successfully created!`);
    window.location.replace(`#/${newPanel.id}`);
  } catch (e) {
    toasts!.addDanger(
      'Error cloning Observability Dashboard, please make sure you have the correct permission.'
    );
    console.error(e);
  }
};

const saveRenamedPanel = async (id, name) => {
  const renamePanelObject = {
    panelId: id,
    panelName: name,
  };

  return http.post(`${CUSTOM_PANELS_API_PREFIX}/panels/rename`, {
    body: JSON.stringify(renamePanelObject),
  });
};

const saveRenamedPanelSO = async (id, name) => {
  const panel: SavedObject<PanelType> = await coreRefs.savedObjectsClient!.get(
    CUSTOM_PANELS_SAVED_OBJECT_TYPE,
    id
  );
  panel.title = name;
  await coreRefs.savedObjectsClient!.update(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id, panel);
};

// Renames an existing CustomPanel
export const renameCustomPanel = (editedCustomPanelName: string, id: string) => async (
  dispatch,
  getState
) => {
  const panel = getState().customPanel.panelList.find((p) => p.id === id);
  const updatedPanel = { ...panel, title: editedCustomPanelName };
  dispatch(
    updatePanel(
      updatedPanel,
      `Operational Panel successfully renamed into "${editedCustomPanelName}"`,
      'Error renaming Operational Panel, please make sure you have the correct permission.'
    )
  );
};

/*
 ** UTILITY FUNCTIONS
 */
const savedObjectToCustomPanel = (so: SimpleSavedObject<PanelType>): CustomPanelType => ({
  id: so.id,
  type: so.type,
  objectId: so.type + ':' + so.id,
  ...so.attributes,
  savedObject: true,
});

const savedObjectPanelsClient = {
  find: (options) =>
    coreRefs.savedObjectsClient!.find({ type: CUSTOM_PANELS_SAVED_OBJECT_TYPE, ...options }),
  delete: (id) => coreRefs.savedObjectsClient!.delete(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id),
  update: (panel) =>
    coreRefs.savedObjectsClient!.update(CUSTOM_PANELS_SAVED_OBJECT_TYPE, panel.id, panel),
  get: (id) => coreRefs.savedObjectsClient!.get(CUSTOM_PANELS_SAVED_OBJECT_TYPE, id),
  create: (panel) => coreRefs.savedObjectsClient!.create(CUSTOM_PANELS_SAVED_OBJECT_TYPE, panel),
};
