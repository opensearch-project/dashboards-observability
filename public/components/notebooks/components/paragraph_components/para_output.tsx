/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCodeBlock, EuiSpacer, EuiText } from '@elastic/eui';
import MarkdownRender from '@nteract/markdown';
import { Media } from '@nteract/outputs';
import moment from 'moment';
import React from 'react';
import { CoreStart } from '../../../../../../../src/core/public';
import {
  DashboardContainerInput,
  DashboardStart,
} from '../../../../../../../src/plugins/dashboard/public';
import { ParaType } from '../../../../../common/types/notebooks';
import { getOSDHttp, getPPLService, uiSettingsService } from '../../../../../common/utils';
import { VisualizationContainer } from '../../../../components/custom_panels/panel_modules/visualization_container';
import PPLService from '../../../../services/requests/ppl';
import { QueryDataGridMemo } from './para_query_grid';

const createQueryColumns = (jsonColumns: any[]) => {
  let index = 0;
  const datagridColumns = [];
  for (index = 0; index < jsonColumns.length; ++index) {
    const datagridColumnObject = {
      id: jsonColumns[index].name,
      displayAsText: jsonColumns[index].name,
    };
    datagridColumns.push(datagridColumnObject);
  }
  return datagridColumns;
};

const getQueryOutputData = (queryObject: any) => {
  const data = [];
  let index = 0;
  let schemaIndex = 0;
  for (index = 0; index < queryObject.datarows.length; ++index) {
    const datarowValue = {};
    for (schemaIndex = 0; schemaIndex < queryObject.schema.length; ++schemaIndex) {
      const columnName = queryObject.schema[schemaIndex].name;
      if (typeof queryObject.datarows[index][schemaIndex] === 'object') {
        datarowValue[columnName] = JSON.stringify(queryObject.datarows[index][schemaIndex]);
      } else if (typeof queryObject.datarows[index][schemaIndex] === 'boolean') {
        datarowValue[columnName] = queryObject.datarows[index][schemaIndex].toString();
      } else {
        datarowValue[columnName] = queryObject.datarows[index][schemaIndex];
      }
    }
    data.push(datarowValue);
  }
  return data;
};

const OutputBody = ({
  key,
  typeOut,
  val,
  para,
  visInput,
  setVisInput,
  DashboardContainerByValueRenderer,
}: {
  key: string;
  typeOut: string;
  val: string;
  para: ParaType;
  visInput: DashboardContainerInput;
  setVisInput: (input: DashboardContainerInput) => void;
  DashboardContainerByValueRenderer: DashboardStart['DashboardContainerByValueRenderer'];
}) => {
  /* Returns a component to render paragraph outputs using the para.typeOut property
   * Currently supports HTML, TABLE, IMG
   * TODO: add table rendering
   */

  const dateFormat = uiSettingsService.get('dateFormat');

  if (typeOut !== undefined) {
    switch (typeOut) {
      case 'QUERY':
        const inputQuery = para.inp.substring(4, para.inp.length);
        const queryObject = JSON.parse(val);
        if (queryObject.hasOwnProperty('error')) {
          return <EuiCodeBlock key={key}>{val}</EuiCodeBlock>;
        } else {
          const columns = createQueryColumns(queryObject.schema);
          const data = getQueryOutputData(queryObject);
          return (
            <div>
              <EuiText key={'query-input-key'} className="wrapAll" data-test-subj="queryOutputText">
                <b>{inputQuery}</b>
              </EuiText>
              <EuiSpacer />
              <QueryDataGridMemo
                key={key}
                rowCount={queryObject.datarows.length}
                queryColumns={columns}
                dataValues={data}
              />
            </div>
          );
        }
      case 'MARKDOWN':
        return (
          <EuiText
            key={key}
            className="wrapAll markdown-output-text"
            data-test-subj="markdownOutputText"
            size="s"
          >
            <MarkdownRender source={val} />
          </EuiText>
        );
      case 'VISUALIZATION':
        let from = moment(visInput?.timeRange?.from).format(dateFormat);
        let to = moment(visInput?.timeRange?.to).format(dateFormat);
        from = from === 'Invalid date' ? visInput.timeRange.from : from;
        to = to === 'Invalid date' ? visInput.timeRange.to : to;
        return (
          <>
            <EuiText size="s" style={{ marginLeft: 9 }}>
              {`${from} - ${to}`}
            </EuiText>
            <DashboardContainerByValueRenderer
              key={key}
              input={visInput}
              onInputUpdated={setVisInput}
            />
          </>
        );
      case 'OBSERVABILITY_VISUALIZATION':
        let fromObs = moment(visInput?.timeRange?.from).format(dateFormat);
        let toObs = moment(visInput?.timeRange?.to).format(dateFormat);
        fromObs = fromObs === 'Invalid date' ? visInput.timeRange.from : fromObs;
        toObs = toObs === 'Invalid date' ? visInput.timeRange.to : toObs;
        const onEditClick = (savedVisualizationId: string) => {
          window.location.assign(`observability-logs#/explorer/${savedVisualizationId}`);
        };
        return (
          <>
            <EuiText size="s" style={{ marginLeft: 9 }}>
              {`${fromObs} - ${toObs}`}
            </EuiText>
            <div style={{ height: '300px', width: '100%' }}>
              <VisualizationContainer
                http={getOSDHttp()}
                editMode={false}
                visualizationId={''}
                onEditClick={onEditClick}
                savedVisualizationId={para.visSavedObjId}
                pplService={getPPLService()}
                fromTime={para.visStartTime}
                toTime={para.visEndTime}
                onRefresh={false}
                pplFilterValue={''}
                usedInNotebooks={true}
              />
            </div>
          </>
        );
      case 'HTML':
        return (
          <EuiText key={key}>
            {/* eslint-disable-next-line react/jsx-pascal-case */}
            <Media.HTML data={val} />
          </EuiText>
        );
      case 'TABLE':
        return <pre key={key}>{val}</pre>;
      case 'IMG':
        return <img alt="" src={'data:image/gif;base64,' + val} key={key} />;
      default:
        return <pre key={key}>{val}</pre>;
    }
  } else {
    console.log('output not supported', typeOut);
    return <pre />;
  }
};

/*
 * "ParaOutput" component is used by notebook to populate paragraph outputs for an open notebook.
 *
 * Props taken in as params are:
 * para - parsed paragraph from notebook
 *
 * Outputs component of nteract used as a container for notebook UI.
 * https://components.nteract.io/#outputs
 */
export const ParaOutput = (props: {
  http: CoreStart['http'];
  pplService: PPLService;
  para: ParaType;
  visInput: DashboardContainerInput;
  setVisInput: (input: DashboardContainerInput) => void;
  DashboardContainerByValueRenderer: DashboardStart['DashboardContainerByValueRenderer'];
}) => {
  const { para, DashboardContainerByValueRenderer, visInput, setVisInput } = props;

  return (
    !para.isOutputHidden && (
      <>
        {para.typeOut.map((typeOut: string, tIdx: number) => {
          return (
            <OutputBody
              key={para.uniqueId + '_paraOutputBody'}
              typeOut={typeOut}
              val={para.out[tIdx]}
              para={para}
              visInput={visInput}
              setVisInput={setVisInput}
              DashboardContainerByValueRenderer={DashboardContainerByValueRenderer}
            />
          );
        })}
      </>
    )
  );
};
