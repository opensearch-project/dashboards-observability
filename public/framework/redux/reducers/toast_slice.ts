import { createSlice } from '@reduxjs/toolkit';

const toastSlice = createSlice({
  name: 'toast',
  initialState: { toasts: [], toastRightSide: true },
  reducers: {
    appendToast: (state, action) => {
      state.toasts = [...state.toasts, action.payload];
    },

    setToastRightSide: (state, action) => (state.toastRightSide = action.payload),

    dismissToast: (state, action) => {
      state.toasts = state.toasts.filter((t) => t.id !== action.payload.id);
    },

    resetToasts: (state, action) => {
      state.toasts = [];
      state.toastRightSide = true;
    },
  },
});

export const toastReducer = toastSlice.reducer;

export const { dismissToast, resetToasts } = toastSlice.actions;
const { appendToast, setToastRightSide } = toastSlice.actions;

export const selectToasts = (rootState) => rootState.toast.toasts;

export const selectToastRightSide = (rootState) => rootState.toast.toastRightSide;

export const addToast = (title, color?, textChild?, side?) => (dispatch, getState) => {
  const newToast = { id: new Date().toISOString(), title, textChild, color };
  dispatch(appendToast(newToast));

  if (side) {
    dispatch(setToastRightSide(side === 'left'));
  }
};
