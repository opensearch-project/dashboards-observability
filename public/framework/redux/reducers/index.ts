/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { combineReducers } from 'redux';

import { queriesReducer } from '../../../components/event_analytics/redux/slices/query_slice';
import { queryResultsReducer } from '../../../components/event_analytics/redux/slices/query_result_slice';
import { queryTabReducer } from '../../../components/event_analytics/redux/slices/query_tab_slice';
import { fieldsReducer } from '../../../components/event_analytics/redux/slices/field_slice';
import { countDistributionReducer } from '../../../components/event_analytics/redux/slices/count_distribution_slice';
import { explorerVisualizationReducer } from '../../../components/event_analytics/redux/slices/visualization_slice';
import { explorerVisualizationConfigReducer } from '../../../components/event_analytics/redux/slices/viualization_config_slice';
import { patternsReducer } from '../../../components/event_analytics/redux/slices/patterns_slice';
import { metricsReducers } from '../../../components/metrics/redux/slices/metrics_slice';
import { panelReducer } from '../../../components/custom_panels/redux/panel_slice';
import { searchMetaDataSliceReducer } from '../../../components/event_analytics/redux/slices/search_meta_data_slice';

const combinedReducer = combineReducers({
  // explorer reducers
  queries: queriesReducer,
  queryResults: queryResultsReducer,
  explorerTabs: queryTabReducer,
  fields: fieldsReducer,
  countDistribution: countDistributionReducer,
  explorerVisualization: explorerVisualizationReducer,
  explorerVisualizationConfig: explorerVisualizationConfigReducer,
  patterns: patternsReducer,
  metrics: metricsReducers,
  customPanel: panelReducer,
  searchMetadata: searchMetaDataSliceReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const rootReducer = combinedReducer;
