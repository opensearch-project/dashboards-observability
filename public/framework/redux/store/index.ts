/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage/session';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import { rootReducer } from '../reducers';

const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const dataStore = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
  enhancers: [],
});

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./rootReducer', () => {
    store.replaceReducer(persistReducer(persistConfig, require('./rootReducer').default)); // eslint-disable-line @typescript-eslint/no-var-requires
  });
}

export type AppDispatch = typeof dataStore.dispatch;

export const persistor = persistStore(dataStore);

export const store = dataStore;
