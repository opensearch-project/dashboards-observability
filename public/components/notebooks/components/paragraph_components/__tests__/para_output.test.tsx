/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render } from '@testing-library/react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { legacy_createStore as createStore } from 'redux';
import {
  getOSDHttp,
  setPPLService,
  uiSettingsService,
} from '../../../../../../common/utils/core_services';
import {
  sampleObservabilityVizParagraph,
  sampleParsedParagraghs1,
} from '../../../../../../test/notebooks_constants';
import { rootReducer } from '../../../../../framework/redux/reducers';
import PPLService from '../../../../../services/requests/ppl';
import { ParaOutput } from '../para_output';

describe('<ParaOutput /> spec', () => {
  configure({ adapter: new Adapter() });
  const store = createStore(rootReducer);

  it('renders markdown outputs', () => {
    const para = sampleParsedParagraghs1[0];
    para.isSelected = true;
    const setVisInput = jest.fn();
    const utils = render(
      <ParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        DashboardContainerByValueRenderer={jest.fn()}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders query outputs', () => {
    const para = sampleParsedParagraghs1[3];
    para.isSelected = true;
    const setVisInput = jest.fn();
    const utils = render(
      <ParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        DashboardContainerByValueRenderer={jest.fn()}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders query outputs with error', () => {
    const para = sampleParsedParagraghs1[3];
    para.out = ['{"error":"Invalid SQL query"}'];
    para.isSelected = true;
    const setVisInput = jest.fn();
    const utils = render(
      <ParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        DashboardContainerByValueRenderer={jest.fn()}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders dashboards visualization outputs', () => {
    const para = sampleParsedParagraghs1[2];
    para.isSelected = true;

    uiSettingsService.get = jest.fn().mockReturnValue('YYYY-MMM-DD HH:mm:ss');
    const setVisInput = jest.fn();
    const utils = render(
      <ParaOutput
        key={para.uniqueId}
        para={para}
        visInput={{
          timeRange: { from: '2020-JUL-21 18:37:44', to: '2020-AUG-20 18:37:44' },
        }}
        setVisInput={setVisInput}
        DashboardContainerByValueRenderer={() => null}
      />
    );
    expect(utils.container.textContent).toMatch('2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44');
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders observability visualization outputs', () => {
    setPPLService(new PPLService(getOSDHttp()));
    const para = sampleObservabilityVizParagraph;
    para.isSelected = true;

    uiSettingsService.get = jest.fn().mockReturnValue('YYYY-MMM-DD HH:mm:ss');
    const setVisInput = jest.fn();
    const utils = render(
      <Provider store={store}>
        <ParaOutput
          key={para.uniqueId}
          para={para}
          visInput={{
            timeRange: { from: '2020-JUL-21 18:37:44', to: '2020-AUG-20 18:37:44' },
          }}
          setVisInput={setVisInput}
          DashboardContainerByValueRenderer={() => null}
        />
      </Provider>
    );
    expect(utils.container.textContent).toMatch('2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44');
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders other types of outputs', () => {
    const para = sampleParsedParagraghs1[0];
    para.isSelected = true;
    para.typeOut = ['HTML', 'TABLE', 'IMG', 'UNKNOWN', undefined];
    para.out = ['', '', '', '', ''];
    const setVisInput = jest.fn();
    const utils = render(
      <ParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        DashboardContainerByValueRenderer={jest.fn()}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
