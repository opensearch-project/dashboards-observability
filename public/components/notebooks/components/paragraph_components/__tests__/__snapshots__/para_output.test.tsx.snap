// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParaOutput /> spec renders dashboards visualization outputs 1`] = `
<div
  class="euiText euiText--small"
  style="margin-left: 9px;"
>
  2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44
</div>
`;

exports[`<ParaOutput /> spec renders markdown outputs 1`] = `
<div
  class="euiText euiText--medium wrapAll markdown-output-text"
>
  <div
    class="markdown-body "
  >
    <h1>
      Type your input here
    </h1>
  </div>
</div>
`;

exports[`<ParaOutput /> spec renders observability visualization outputs 1`] = `
<div
  class="euiText euiText--small"
  style="margin-left: 9px;"
>
  2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44
</div>
`;

exports[`<ParaOutput /> spec renders other types of outputs 1`] = `
<div
  class="euiText euiText--medium"
>
  <div
    class="sc-csCMJt hFpeNP"
  />
</div>
`;

exports[`<ParaOutput /> spec renders query outputs 1`] = `
<div>
  <div
    class="euiText euiText--medium wrapAll"
  >
    <b>
       select * from opensearch_dashboards_sample_data_flights limit 2
    </b>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    id="queryDataGrid"
  >
    <span
      class="euiLoadingSpinner euiLoadingSpinner--xLarge"
    />
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="-1"
    />
    <div
      class="euiDataGrid__focusWrap"
      data-focus-lock-disabled="disabled"
    >
      <div
        class="euiDataGrid euiDataGrid--bordersAll euiDataGrid--headerShade euiDataGrid--footerOverline euiDataGrid--rowHoverHighlight euiDataGrid--stickyFooter"
      />
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`<ParaOutput /> spec renders query outputs with error 1`] = `
<div
  class="euiCodeBlock euiCodeBlock--fontSmall euiCodeBlock--paddingLarge prismjs language-none"
>
  <pre
    class="euiCodeBlock__pre euiCodeBlock__pre--whiteSpacePreWrap"
    tabindex="-1"
  >
    <code
      class="euiCodeBlock__code"
    >
      {"error":"Invalid SQL query"}
    </code>
  </pre>
</div>
`;
